File: src/player.c
/*
The MIT License (MIT)

Copyright (c) 2014 Master.G

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

#include "player.h"
#include "game.h"
#include "standard_ai.h"
#include "advanced_ai.h"

void Player_SetupStandardAI(player_t *player) {
  player->eventHandlers[Player_Event_GetReady] = StandardAI_GetReady;
  player->eventHandlers[Player_Event_Bid] = StandardAI_Bid;
  player->eventHandlers[Player_Event_Play] = StandardAI_Play;
  player->eventHandlers[Player_Event_Beat] = StandardAI_Beat;
}

void Player_SetupAdvancedAI(player_t *player) {
  player->eventHandlers[Player_Event_GetReady] = AdvancedAI_GetReady;
  player->eventHandlers[Player_Event_Bid] = StandardAI_Bid;
  player->eventHandlers[Player_Event_Play] = StandardAI_Play;
  player->eventHandlers[Player_Event_Beat] = AdvancedAI_Beat;
}

void Player_Destroy(player_t *player) {
  HandList_Destroy(&player->handlist);
  free(player);
}

void Player_Clear(player_t *player) {
  HandList_Destroy(&player->handlist);
  player->handlist = NULL;
  player->identity = PlayerIdentity_Peasant;
  CardArray_Clear(&player->cards);
  CardArray_Clear(&player->record);
}

int Player_HandleEvent(void *p, int event, void *game) {
  player_t *player = (player_t *) p;

  return player->eventHandlers[event](p, game);
}

File: lualandlord/player.lua
require "lualandlord/hand";
require "lualandlord/standard_ai";
require "lualandlord/advanced_ai";

ll.PlayerEvent_GetReady 	= 1;
ll.PlayerEvent_Bid 			= 2;
ll.PlayerEvent_Start		= 3;
ll.PlayerEvent_Play			= 4;
ll.PlayerEvent_Beat 		= 5;

ll.PlayerIdentity_Peasant	= 1;
ll.PlayerIdentity_Landlord	= 2;

ll.PlayerBid_Abstain		= 1;
ll.PlayerBid_Bid			= 2;


-----------------------------------------------------------------------------
-- player
--
-- {
-- 		cards			: CardArray
-- 		record			: CardArray
--		handlist		: {}
--		identity 		: number
-- 		seatId 			: number
-- 		bid 			: number
-- 		eventHandlers	: {}
-- }
-----------------------------------------------------------------------------

-- create a empty player
function ll.Player_Create()
	local player = {};
	player.cards = ll.CardArray_Create();
	player.record = ll.CardArray_Create();
	player.handlist = {};
	player.seatId = 0;
	player.bid = 0;
	player.eventHandlers = {};

	return player;
end

-- create a standard AI player
function ll.Player_CreateStandardAI()
	local player = ll.Player_Create();

	player.eventHandlers[ll.PlayerEvent_GetReady] 	= ll.StandardAI_GetReady;
	player.eventHandlers[ll.PlayerEvent_Bid] 		= ll.StandardAI_Bid;
	player.eventHandlers[ll.PlayerEvent_Play] 		= ll.StandardAI_Play;
	player.eventHandlers[ll.PlayerEvent_Beat]		= ll.StandardAI_Beat;

	return player;
end

-- create a advanced AI player
function ll.Player_CreateAdvancedAI()
	local player = ll.Player_Create();

	player.eventHandlers[ll.PlayerEvent_GetReady] 	= ll.AdvancedAI_GetReady;
	player.eventHandlers[ll.PlayerEvent_Bid] 		= ll.StandardAI_Bid;
	player.eventHandlers[ll.PlayerEvent_Play] 		= ll.StandardAI_Play;
	player.eventHandlers[ll.PlayerEvent_Beat]		= ll.AdvancedAI_Beat;

	return player;
end

-- clear a player's context
function ll.Player_Clear(player)
	player.handlist = {};
	player.identity = ll.PlayerIdentity_Peasant;
	ll.CardArray_Clear(player.cards);
	ll.CardArray_Clear(player.record);
end

-- handle event
function ll.Player_HandleEvent(player, event, ctx)
	return player.eventHandlers[event](player, ctx);
end

